General dependency upgrades (awx venv only for now)

AlanCoding opened the pull request:
##### SUMMARY
This implements a process change and does a general upgrade of everything in the AWX virtual environment which won't break things.

Like all dependency upgrades, most of the real work involves categorizing which things will break things on upgrade. To do this in a more maintainable way, I've written blurbs on those things we can't upgrade, and then _unpinned_ the content in `requirements.in` for those that don't break things.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
9.0.1
```


##### ADDITIONAL INFORMATION

This needs some corresponding changes to unblock upgrade of pip & setuptools, but I have been testing it with temporary hacks for that and everything is going fine, so it looks like the ecosystem has worked out enough to not error here.

fosterseth committed:
Update pip and setuptools in requirements txt

Versions selected to be pre-19 pip
due to unresolved issues with the build systems

Upgrade everything, party on

document new process

rotate license files

fix Swagger schema generation target

Remove --ignore-installed flag

AlanCoding commented:
New lib license links:

https://pypi.org/project/zipp/
https://github.com/jaraco/zipp/blob/master/LICENSE

https://github.com/tobgu/pyrsistent/blob/master/LICENCE.mit
https://pypi.org/project/pyrsistent/

https://bitbucket.org/ruamel/yaml.clib/src/default/LICENSE
https://pypi.org/project/ruamel.yaml.clib/

https://github.com/tobgu/pyrsistent/blob/master/LICENCE.mit
https://pypi.org/project/pyrsistent/

https://gitlab.com/python-devs/importlib_resources/blob/master/LICENSE
https://pypi.org/project/importlib_resources/

AlanCoding commented:
Looks like the swagger failure was introduced by an upgrade of DRF in this between the two versions

https://github.com/encode/django-rest-framework/compare/3.9.4...3.10.0

and I know that it was introduced _before_ 7762aaa90f92ff30e521931db2ec65c52f8da98e, still searching locally.

AlanCoding commented:
passes with

https://github.com/encode/django-rest-framework/commit/9811a29a5a1348f1a5de30a9a7a5d0f8d2fd4843

fails with

https://github.com/encode/django-rest-framework/commit/c3a13916d2c8025b042a9b9fb1ec5af25f202cb4

I think I'm confusing myself because they did branch merges, and I'm only interested in the 3.10 branch

AlanCoding commented:
introduced in https://github.com/encode/django-rest-framework/commit/37f210a455cc92cb3f61a23e194a1d0de58d149b

> Added OpenAPI Schema Generation.

https://github.com/encode/django-rest-framework/pull/6532

> SchemaView and in-built interactive docs will (likely) require this to be set back to `rest_framework.schemas.AutoSchema` to function correctly.

AlanCoding commented:
Don't want to lose this

https://github.com/ansible/ansible/pull/62895

shanemcd committed:
Bootstrap venv creation with flit and poetry

this is in addition to pip and setuptools installs

add in --ignore-installed and --no-build-isolation flags
reasoning is that we are not installing needed packages
for building

AlanCoding committed:
Avoid system site packages and add more setup things

AlanCoding committed:
Back off setuptools version due to allow_hosts option error

ryanpetrello commented on requirements/requirements.txt:
```diff
@@ -1,135 +1,132 @@
-django==2.2.4
+django-split-settings==1.0.0
+django-taggit==1.2.0
+django==2.2.7
```
👀 https://github.com/advisories/GHSA-hvmf-r92r-27hr

(we should go with 2.2.8)

AlanCoding committed:
Bump Django version which recently updated

AlanCoding committed:
Attempts to simplify build environment

Remove build isolation flag

Do not use --ignore-installed for any pip install commands

Add Makefile comments

do not use system site packages for awx venv

Consolidate bootstrap pins

Do another upgrade, properly document wheel

AlanCoding commented on Makefile:
```diff
@@ -188,12 +190,13 @@ requirements_ansible_dev:
 	fi
 	echo "include-system-site-packages = true" >> $(VENV_BASE)/awx/lib/python$(PYTHON_VERSION)/pyvenv.cfg
 	$(VENV_BASE)/awx/bin/pip uninstall --yes -r requirements/requirements_tower_uninstall.txt
+	rm $(shell ls -d $(VENV_BASE)/awx/lib/python* | head -n 1)/no-global-site-packages.txt
```
Still plan to document what this is doing with some comments

AlanCoding commented:
May need to revisit the version of setuptools, or else mark as an upgrade blocker. We hit a complex versioning issue related to https://github.com/pypa/setuptools/issues/1916

AlanCoding committed:
Get rid of hacks to simplify process

Remove poetry and wheel from bootstrap installs

Cleanup some docs around them

AlanCoding committed:
update paramiko source

AlanCoding commented:
recheck

AlanCoding committed:
Fallback to globals for these tests

AlanCoding commented:
recheck
